### PR TITLE
Make Service Bus attempt to reconnect forever in case of issues

### LINKED ISSUE #1612: Disable service bus subscription (and potentially other errors?) will stop message processing in Dapr sidecar and needs pod restart to resume
## In what area(s)?
/area runtime
## What version of Dapr?
1.6.0
## Expected Behavior
Intermittent service bus errors should not stop Dapr message processing
## Actual Behavior
We discovered this issue when disabling the service bus subscription to stop message processing. We immediately get errors in the dapr side car on the form:
```
time="2022-03-24T18:24:01.222738335Z" level=debug msg="Closing subscription to topic <topic>" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
time="2022-03-24T18:24:01.223704688Z" level=error msg="azure service bus error: error receiving message on topic <topic>, link detached, reason: *Error{Condition: amqp:link:detach-forced, Description: The link 'G3:2832318:b7gVHOmwF2ISmLz21XWy-R0wQPd4AWj7JbcEyHzxEVVqJiw_nyg6HA' is force detached. Code: consumer(link2688). Details: InnerMessageReceiver was closed. TrackingId:c1d38a450005056a00000a80623c6ae1_G3_B2, SystemTracker:<service bus namespace>:Topic:<topic>|<subscription>, Timestamp:2022-03-24T18:24:01, Info: map[]}" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
time="2022-03-24T18:24:01.223755891Z" level=debug msg="Lock renewal context for topic <topic> done" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
time="2022-03-24T18:24:01.223667186Z" level=warning msg="azure service bus error: closing subscription entity for topic <topic>: link detached, reason: *Error{Condition: amqp:link:detach-forced, Description: The link 'G3:2832318:b7gVHOmwF2ISmLz21XWy-R0wQPd4AWj7JbcEyHzxEVVqJiw_nyg6HA' is force detached. Code: consumer(link2688). Details: InnerMessageReceiver was closed. TrackingId:c1d38a450005056a00000a80623c6ae1_G3_B2, SystemTracker:<service bus namespace>:Topic:<topic>|<subscription>, Timestamp:2022-03-24T18:24:01, Info: map[]}" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
time="2022-03-24T18:24:01.22373459Z" level=warning msg="Subscription to topic <topic> lost connection, attempting to reconnect... [0/30]" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
time="2022-03-24T18:24:03.224120298Z" level=debug msg="Subscription to topic <topic> is limited to 4 message handler(s)" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
time="2022-03-24T18:24:03.224174001Z" level=debug msg="Waiting to receive message on topic <topic>" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
time="2022-03-24T18:24:03.984272457Z" level=error msg="azure service bus error: error receiving message on topic <topic>, *Error{Condition: com.microsoft:entity-disabled, Description: Messaging entity '<service bus namespace>:Topic:<topic>|<subscription>' is currently disabled. For more information please see https://aka.ms/ServiceBusExceptions . TrackingId:f3befc0e-acbf-42f3-8e7d-73316da0f357_B2, SystemTracker:<service bus namespace>:Topic:<topic>|<subscription>, Timestamp:2022-03-24T18:24:03 TrackingId:9d8eea867acd4c45b77edf5128205c1e_G4, SystemTracker:gateway5, Timestamp:2022-03-24T18:24:03, Info: map[]}" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
```
And after only 1 minute and 20 seconds we get this error: 
```
time="2022-03-24T18:25:20.086919138Z" level=debug msg="Waiting to receive message on topic <topic>" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
time="2022-03-24T18:25:20.086849334Z" level=debug msg="Subscription to topic <topic> is limited to 4 message handler(s)" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
time="2022-03-24T18:25:20.676660156Z" level=debug msg="Closing subscription to topic <topic>" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
time="2022-03-24T18:25:20.676746161Z" level=debug msg="Reconnect context for topic <topic> is done" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
time="2022-03-24T18:25:20.676706759Z" level=error msg="azure service bus error: error receiving message on topic <topic>, *Error{Condition: com.microsoft:entity-disabled, Description: Messaging entity '<service bus namespace>:Topic:<topic>|<subscription>' is currently disabled. For more information please see https://aka.ms/ServiceBusExceptions . TrackingId:f3befc0e-acbf-42f3-8e7d-73316da0f357_B2, SystemTracker:<service bus namespace>:Topic:<topic>|<subscription>, Timestamp:2022-03-24T18:25:20 TrackingId:f597010729174588bb5d37083523fffa_G6, SystemTracker:gateway5, Timestamp:2022-03-24T18:25:20, Info: map[]}" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
time="2022-03-24T18:25:20.676824965Z" level=debug msg="Lock renewal context for topic <topic> done" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
time="2022-03-24T18:25:20.67672946Z" level=error msg="Subscription to topic <topic> lost connection, unable to recover after 30 attempts" app_id=<deployment> instance=<pod name> scope=dapr.contrib type=log ver=1.6.0
```
After this point, no more messages are processed by the pod, and we need to restart the pod to the the processing going again.
## Steps to Reproduce the Problem
- Create service that subscribes to service bus subscription
- Deploy service in Kubernetes
- In Azure portal, disable subscription
- Observe similar logs in dapr sidecar.
- It is unclear what other errors in service bus that will cause this. In any case, 30 retries in 1 minute and 20 seconds seems to be giving up too fast, without the possibility to "self heal".
## Release Note
RELEASE NOTE: **FIX** Self healing from intermittent issues in Azure Service Bus

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
# Description

With exponential back-off configurable between min and max time. Fixes #1612

Also includes fixes:

- Binding: make sure it actually retries to connect forever
- Binding: add delay (exponential backoff) before reconnecting
- PubSub: better handling of failures such as topics disabled or other non-connection issues

## Issue reference

Fixes #1612

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Extended the documentation https://github.com/dapr/docs/pull/2512

### CODE REVIEW

File: pubsub/azure/servicebus/metadata.go [hunk 1]
```diff
@@ -22,8 +22,8 @@ type metadata struct {
 	HandlerTimeoutInSec             int    `json:"handlerTimeoutInSec"`
 	LockRenewalInSec                int    `json:"lockRenewalInSec"`
 	MaxActiveMessages               int    `json:"maxActiveMessages"`
-	MaxReconnectionAttempts         int    `json:"maxReconnectionAttempts"`
-	ConnectionRecoveryInSec         int    `json:"connectionRecoveryInSec"`
+	MaxConnectionRecoveryInSec      int    `json:"maxConnectionRecoveryInSec"`
```
Comment (berndverst) [inline pubsub/azure/servicebus/metadata.go:25]: This is a breaking change - see https://docs.dapr.io/reference/components-reference/supported-pubsub/setup-azure-servicebus/

**Review by berndverst — CHANGES_REQUESTED** [review]
This is a breaking change as the metadata options has changed and existing metadata options no longer take effect. That needs to be handled.

**Comment by ItalyPaleAle** [comment]
It sort if is, and sort of isn't.

The additional metadata keys are ignored so that won't cause crashes.

Now, it's true that it changes the behavior, but I don't believe we should be particularly concerned with that, because the current behavior is incorrect/undesirable.

Per linked issue (user-reported), the pubsub component would only try to reconnect up to `maxReconnectionAttempts`. After that, it would enter in a state in which nothing is happening. The runtime doesn't crash, but it doesn't receive any message. This is arguably not a desirable behavior and it should be ok to remove it.

For `connectionRecoveryInSec`, with the change to make the component retry forever, it makes sense to have a min and max instead. Although users that have set this value before may see it ignored and see a different behavior, it shouldn't be a significant impact, as it just regulates the reconnection interval (and before it was only used up to the `maxReconnectionAttempts`).

**Comment by berndverst** [comment]
Behavior change is a breaking change @ItalyPaleAle. Of course I know that metadata keys which don't exist are ignored.

I wonder if there is any situation where a user previously benefited from retries and would not be broken as a result. Is there something better we can do here?

Maybe at a minimum we just write a log message that the metadata option is ignored and other metadata options (the new ones you have added) should be used instead?

**Comment by ItalyPaleAle** [comment]
> Maybe at a minimum we just write a log message that the metadata option is ignored and other metadata options (the new ones you have added) should be used instead?

that's a good suggestion

**Comment by ItalyPaleAle** [comment]
Changed as requested

**Review by berndverst — APPROVED** [review]

**Pull request merged**